### ISSUE TITLE
Array not assignable to ReadonlyArray with subclass items

### ISSUE THREAD

**Issue by stephanedr** [Microsoft/TypeScript#10368]
**TypeScript Version:**  2.0.0

**Code**
```ts
class A {
    a: string;
}
class B extends A {
    b: string;
}
function f(a: ReadonlyArray<A>): void { }
f([new A()]);           // OK
f([new A(), new B()]);  // OK
f([new B()]);           // Error
```

**Expected behavior:**
To compile.

**Actual behavior:** 
error TS2345: Argument of type 'B[]' is not assignable to parameter of type 'ReadonlyArray<A>'.
  Types of property 'concat' are incompatible.
    Type '(...items: (B | B[])[]) => B[]' is not assignable to type '{ <U extends ReadonlyArray<A>>(...items: U[]): A[]; (...items: A[]): A[]; }'.
      Types of parameters 'items' and 'items' are incompatible.
        Type 'A' is not assignable to type 'B | B[]'.
          Type 'A' is not assignable to type 'B[]'.
            Property 'length' is missing in type 'A'.

**Comment by sandersn** [comment]
This sounds related to the `concat` fix in #9997. It may have another problem, or it may need to be propagated to `ReadonlyArray` as well.

Have you tried with `typescript@next`?

**Comment by stephanedr** [comment]
Just checked with 2.1.0-dev.20160816. Same error.

**Comment by sandersn** [comment]
I copied the improved signatures from `concat` and it works now. I'll create a PR for the fix.

**Comment by stephanedr** [comment]
Confirmed fixed (2.1.0-dev.20160818). Thanks.
